### ISSUE TITLE
Memory allocation when using multiple platforms/backends

### ISSUE THREAD

**Issue by francescoclarifai** [triton-inference-server/server#2535]
Is inference using multiple platforms/backends within the same server officially supported? How is memory allocation handled?

I'm seeing the libtorch and tensorflow backends competing for GPU memory, eventually resulting in OOM for requests that would otherwise easily fit in the GPU.

Here are the details of the experiment:
- 1000 inference requests with pretrained ResNet50 in both Pytorch and TF.
- Triton-server using 12GB GPU.

Test 1: batch size 1
Everything works ok: 1) sending first Pytorch inference requests, then TF; 2) first TF, then Pytorch; 3) sending the two types of requests simultaneously from two separate clients.
Processing the Pytorch request allocates ~1.3GB; the GPU memory stays allocated once the request has been processed. Subsequent Pytorch requests do not increase allocated memory.
TF run allocates ~8.4GB, the GPU memory stays allocated. Subsequent TF requests do not increase allocated memory.
After running both TF and Pytorch, ~8.6GB are and stay allocated. 

Test 2: batch size 64
Pytorch run allocates ~6GB. When running TF afterwards, it successfully returns responses, but it throws the following warning:
`W tensorflow/core/common_runtime/bfc_allocator.cc:305] Garbage collection: deallocate free memory regions (i.e., allocations) so that we can re-allocate a larger region to avoid OOM due to memory fragmentation. If you see this message frequently, you are running near the threshold of the available device memory and re-allocation may incur great performance overhead. You may try smaller batch sizes to observe the performance impact. Set TF_ENABLE_GPU_GARBAGE_COLLECTION=false if you'd like to disable this feature.
`

Test 3: batch size 100
Single Pytorch run allocates ~10GB. Sending multiple simultaneous Pytorch requests does not increase memory allocation.
Singe TF run allocates ~9GB. Sending multiple simultaneous TF requests does not increase memory allocation.
Sending a single TF request after a Pytorch requests (or viceversa) results in OOM.

In conclusion, it appears whichever platform/backend gets the memory allocation first, won't ever release it until the server is restarted. Subsequent inference request using a different platform will fail with OOM.

Is this the expected behavior? Are there options or future plans to have a common memory allocator at the server level? Or is the usage of multiple platforms on the same server not supported?

**Comment by deadeyegoodwin** [comment]
Models from multiple FWs are supported, as you've shown yourself in your examples. The problem is that each framework has its own GPU memory allocator and the frameworks in general do not provide a way to override that allocator. It is known that TF has an allocator that does not return GPU memory to the system but instead just manages/reuses it. The "high-water" allocation behavior is something that Triton cannot control though we have requested TF to provide an option to return GPU memory to the system. The PyTorch framework appears to have similar behavior.

The model analyzer is a new Triton project that helps you understand the maximum memory usage of your model under the model configuration settings you choose (batching and model instances primarily) so that you can understand if two models can coexist: https://github.com/triton-inference-server/model_analyzer

But it is fair to say that because of this FW behavior you may get the most out of your GPU memory if you group PyT models and TF models on different GPUs.... depending on your use case.
